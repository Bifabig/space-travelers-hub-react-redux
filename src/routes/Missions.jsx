import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Table from 'react-bootstrap/Table';
import Button from 'react-bootstrap/Button';
import Badge from 'react-bootstrap/Badge';
import {
  getMissionItems,
  joinMission,
  leaveMission,
} from '../redux/missions/missionsSlice';
import styles from '../styles/Missions.module.css';

const Missions = () => {
  const dispatch = useDispatch();
  const { missions, isLoading, error } = useSelector((state) => state.missions);
  useEffect(() => {
    if (missions.length === 0) {
      dispatch(getMissionItems());
    }
  }, [dispatch, missions.length]);
  const handleJoinMission = (id) => {
    dispatch(joinMission(id));
  };

  const handleLeaveMission = (id) => {
    dispatch(leaveMission(id));
  };

  if (error) return <h2>Something went wrong</h2>;

  return isLoading ? (
    <h1>Loading...</h1>
  ) : (
    <section className={styles.wrapper}>
      <hr />
      <div className={styles.table}>
        <Table striped bordered hover responsive>
          <thead>
            <tr>
              <th>Mission</th>
              <th>Description</th>
              <th>Status</th>
              <th>{'    '}</th>
            </tr>
          </thead>
          <tbody>
            {missions.map((mission) => (
              <tr key={mission.mission_id}>
                <td>{mission.mission_name}</td>
                <td className={styles.desc}>{mission.description}</td>
                <td>
                  <div className={styles.center}>
                    {mission.reserved ? (
                      <Badge bg="info">Active Member</Badge>
                    ) : (
                      <Badge bg="secondary">Not A Member</Badge>
                    )}
                  </div>
                </td>
                <td>
                  <div className={styles.centerBtn}>
                    {mission.reserved ? (
                      <Button
                        variant="outline-danger"
                        onClick={() => handleLeaveMission(mission.mission_id)}
                      >
                        Leave Mission
                      </Button>
                    ) : (
                      <Button
                        variant="outline-secondary"
                        onClick={() => handleJoinMission(mission.mission_id)}
                      >
                        Join Mission
                      </Button>
                    )}
                  </div>
                </td>
              </tr>
            ))}
          </tbody>
        </Table>
      </div>
    </section>
  );
};

export default Missions;
