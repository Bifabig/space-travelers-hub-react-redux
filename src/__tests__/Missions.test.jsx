import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { useDispatch, useSelector } from 'react-redux';
import Missions from '../routes/Missions';
import {
  getMissionItems,
  joinMission,
  leaveMission,
} from '../components/redux/missions/missionsSlice';
import '@testing-library/jest-dom/extend-expect';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

jest.mock('../components/redux/missions/missionsSlice', () => ({
  getMissionItems: jest.fn(),
  joinMission: jest.fn(),
  leaveMission: jest.fn(),
}));

describe('Missions', () => {
  beforeEach(() => {
    useDispatch.mockReturnValue(jest.fn());
    useSelector.mockReturnValue({
      missions: [],
      isLoading: false,
      error: null,
    });
  });

  it('should render loading message when isLoading is true', () => {
    useSelector.mockReturnValue({
      missions: [],
      isLoading: true,
      error: null,
    });

    render(<Missions />);

    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });

  it('should render error message when error exists', () => {
    useSelector.mockReturnValue({
      missions: [],
      isLoading: false,
      error: 'Something went wrong',
    });

    render(<Missions />);

    expect(screen.getByText('Something went wrong')).toBeInTheDocument();
    const { container } = render(<Missions />);
    expect(container).toMatchSnapshot();
  });

  it('should dispatch getMissionItems when rockets array is empty', () => {
    useSelector.mockReturnValue({
      missions: [],
      isLoading: false,
      error: null,
    });

    const { container } = render(<Missions />);

    expect(getMissionItems).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });

  it('should render the missions', () => {
    useSelector.mockReturnValue({
      missions: [
        {
          mission_id: '1',
          mission_name: 'Thaicom',
          description: 'Mission description',
          reserved: false,
        },
      ],
      isLoading: false,
      error: null,
    });

    const { container } = render(<Missions />);

    expect(screen.getByText('Thaicom')).toBeInTheDocument();
    expect(screen.getByText('Mission description')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('should dispatch joinMission when clicking on "Join Mission" button', () => {
    const missionId = '1';
    useSelector.mockReturnValue({
      missions: [
        {
          mission_id: missionId,
          mission_name: 'Thaicom',
          description: 'Mission description',
          reserved: false,
        },
      ],
      isLoading: false,
      error: null,
    });

    render(<Missions />);

    const reserveButton = screen.getByText('Join Mission');
    fireEvent.click(reserveButton);

    expect(joinMission).toHaveBeenCalledWith(missionId);
  });

  it('should dispatch leaveMission when clicking on "Leave Mission" button', () => {
    const missionId = '1';
    useSelector.mockReturnValue({
      missions: [
        {
          mission_id: missionId,
          mission_name: 'Thaicom',
          description: 'Mission description',
          reserved: true,
        },
      ],
      isLoading: false,
      error: null,
    });

    render(<Missions />);

    const cancelButton = screen.getByText('Leave Mission');
    fireEvent.click(cancelButton);

    expect(leaveMission).toHaveBeenCalledWith(missionId);
  });
});
