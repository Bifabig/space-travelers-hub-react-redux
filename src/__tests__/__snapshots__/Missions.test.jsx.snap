// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions should dispatch getMissionItems when rockets array is empty 1`] = `
<div>
  <section
    class="wrapper"
  >
    <hr />
    <div
      class="table"
    >
      <div
        class="table-responsive"
      >
        <table
          class="table table-striped table-bordered table-hover"
        >
          <thead>
            <tr>
              <th>
                Mission
              </th>
              <th>
                Description
              </th>
              <th>
                Status
              </th>
              <th>
                    
              </th>
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
    </div>
  </section>
</div>
`;

exports[`Missions should render error message when error exists 1`] = `
<div>
  <h2>
    Something went wrong
  </h2>
</div>
`;

exports[`Missions should render the missions 1`] = `
<div>
  <section
    class="wrapper"
  >
    <hr />
    <div
      class="table"
    >
      <div
        class="table-responsive"
      >
        <table
          class="table table-striped table-bordered table-hover"
        >
          <thead>
            <tr>
              <th>
                Mission
              </th>
              <th>
                Description
              </th>
              <th>
                Status
              </th>
              <th>
                    
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                Thaicom
              </td>
              <td
                class="desc"
              >
                Mission description
              </td>
              <td>
                <div
                  class="center"
                >
                  <span
                    class="badge bg-secondary"
                  >
                    Not A Member
                  </span>
                </div>
              </td>
              <td>
                <div
                  class="centerBtn"
                >
                  <button
                    class="btn btn-outline-secondary"
                    type="button"
                  >
                    Join Mission
                  </button>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </section>
</div>
`;
